File: backend/src/main/java/com/breadfactory/erp/service/PricingService.java
package com.breadfactory.erp.service;

import com.breadfactory.erp.entity.Product;
import com.breadfactory.erp.entity.Shop;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.math.RoundingMode;

/**
 * Centralized Pricing Engine & Business Rules Service.
 * Implements buyer-type based pricing calculations and strict minimum selling price validations.
 */
@Service
public class PricingService {

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class PriceCalculationResult {
        private Long productId;
        private String productCode;
        private String productName;
        private String customerType; // SHOP, WHOLESALE_DEALER, CUSTOMER
        private BigDecimal mrp;
        private BigDecimal minimumSellingPrice;
        private BigDecimal discountPercent;
        private BigDecimal basePrice;
        private BigDecimal unitSellingPrice;
        private BigDecimal totalPrice;
        private Integer quantity;
        private boolean valid;
        private String validationMessage;
    }

    /**
     * Calculates the unit selling price and line total according to buyer type rules.
     */
    public PriceCalculationResult calculatePrice(Product product, Shop buyer, BigDecimal requestedPrice, Integer quantity) {
        if (product == null) {
            throw new IllegalArgumentException("Product cannot be null for price calculation.");
        }
        int qty = (quantity != null && quantity > 0) ? quantity : 1;
        String rawType = buyer != null && buyer.getCustomerType() != null ? buyer.getCustomerType().toUpperCase() : "SHOP";

        // Normalize customer type
        String custType;
        if (rawType.contains("WHOLESALE") || rawType.contains("DEALER") || rawType.contains("AGENT")) {
            custType = "WHOLESALE_DEALER";
        } else if (rawType.contains("CUSTOMER") || rawType.contains("RETAIL_CUSTOMER")) {
            custType = "CUSTOMER";
        } else {
            custType = "SHOP";
        }

        BigDecimal mrp = product.getMrp() != null ? product.getMrp() : BigDecimal.valueOf(50.00);
        BigDecimal minSellingPrice = product.getMinimumSellingPrice() != null 
                ? product.getMinimumSellingPrice() 
                : (product.getDealerPrice() != null ? product.getDealerPrice() : mrp);

        PriceCalculationResult.PriceCalculationResultBuilder builder = PriceCalculationResult.builder()
                .productId(product.getId())
                .productCode(product.getProductCode())
                .productName(product.getName())
                .customerType(custType)
                .mrp(mrp)
                .minimumSellingPrice(minSellingPrice)
                .quantity(qty);

        // ── 1. WHOLESALE DEALER PRICING ──────────────────────────────────────────────
        // System calculates individual discount percentage and derives actual unit selling amount
        if ("WHOLESALE_DEALER".equals(custType)) {
            BigDecimal dealerDiscount = buyer != null && buyer.getDiscountPercent() != null 
                    ? buyer.getDiscountPercent() 
                    : BigDecimal.valueOf(10.00); // default 10% if not configured

            BigDecimal base = mrp;
            BigDecimal discountFraction = dealerDiscount.divide(BigDecimal.valueOf(100), 4, RoundingMode.HALF_UP);
            BigDecimal discountAmountPerUnit = base.multiply(discountFraction).setScale(2, RoundingMode.HALF_UP);
            BigDecimal finalUnitPrice = base.subtract(discountAmountPerUnit).setScale(2, RoundingMode.HALF_UP);
            BigDecimal total = finalUnitPrice.multiply(BigDecimal.valueOf(qty)).setScale(2, RoundingMode.HALF_UP);

            return builder
                    .discountPercent(dealerDiscount)
                    .basePrice(base)
                    .unitSellingPrice(finalUnitPrice)
                    .totalPrice(total)
                    .valid(true)
                    .build();
        }

        // ── 2. SHOP PRICING ──────────────────────────────────────────────────────────
        // Enforces minimum selling price rule: selling price >= minimumSellingPrice
        if ("SHOP".equals(custType)) {
            BigDecimal unitPrice = requestedPrice != null ? requestedPrice : minSellingPrice;

            if (unitPrice.compareTo(minSellingPrice) < 0) {
                return builder
                        .unitSellingPrice(unitPrice)
                        .totalPrice(unitPrice.multiply(BigDecimal.valueOf(qty)).setScale(2, RoundingMode.HALF_UP))
                        .valid(false)
                        .validationMessage("Minimum selling price for this product is ₹" + minSellingPrice.setScale(2, RoundingMode.HALF_UP) + ".")
                        .build();
            }

            BigDecimal total = unitPrice.multiply(BigDecimal.valueOf(qty)).setScale(2, RoundingMode.HALF_UP);
            return builder
                    .unitSellingPrice(unitPrice.setScale(2, RoundingMode.HALF_UP))
                    .totalPrice(total)
                    .valid(true)
                    .build();
        }

        // ── 3. CUSTOMER PRICING ──────────────────────────────────────────────────────
        // Manual price entry directly from user
        BigDecimal custPrice = requestedPrice != null 
                ? requestedPrice 
                : (product.getRetailPrice() != null ? product.getRetailPrice() : mrp);

        BigDecimal total = custPrice.multiply(BigDecimal.valueOf(qty)).setScale(2, RoundingMode.HALF_UP);
        return builder
                .unitSellingPrice(custPrice.setScale(2, RoundingMode.HALF_UP))
                .totalPrice(total)
                .valid(true)
                .build();
    }
}

File: admin-dashboard/src/services/pricingService.ts
/**
 * Centralized Pricing Engine & Business Rules Service for Frontend.
 * Enforces:
 * 1. Shop: Minimum Selling Price rule (Price >= minimumSellingPrice)
 * 2. Wholesale Dealer: Automatic calculation from dealer's configured discount % (Displays actual selling amount)
 * 3. Customer: Manual price entry during sales, no onboarding pricing configuration
 */

export interface PricingProduct {
  id: number;
  productCode?: string;
  name: string;
  mrp: number;
  minimumSellingPrice?: number;
  dealerPrice?: number;
  wholesalePrice?: number;
  retailPrice?: number;
}

export interface PricingBuyer {
  id?: number;
  name?: string;
  customerType?: 'SHOP' | 'WHOLESALE_AGENT' | 'WHOLESALE_DEALER' | 'RETAIL_CUSTOMER' | 'CUSTOMER' | string;
  discountPercent?: number;
}

export interface CalculatedItemPrice {
  productId: number;
  productName: string;
  customerType: 'SHOP' | 'WHOLESALE_DEALER' | 'CUSTOMER';
  mrp: number;
  minimumSellingPrice: number;
  discountPercent: number;
  basePrice: number;
  unitSellingPrice: number;
  totalPrice: number;
  quantity: number;
  isValid: boolean;
  validationMessage?: string;
}

export const normalizeCustomerType = (type?: string): 'SHOP' | 'WHOLESALE_DEALER' | 'CUSTOMER' => {
  if (!type) return 'SHOP';
  const t = type.toUpperCase();
  if (t.includes('WHOLESALE') || t.includes('DEALER') || t.includes('AGENT')) return 'WHOLESALE_DEALER';
  if (t.includes('CUSTOMER') || t.includes('RETAIL')) return 'CUSTOMER';
  return 'SHOP';
};

/**
 * Calculates selling price and validates buyer rules.
 */
export const calculateProductSellingPrice = (
  product: PricingProduct,
  buyer?: PricingBuyer,
  requestedPrice?: number,
  quantity: number = 1
): CalculatedItemPrice => {
  const custType = normalizeCustomerType(buyer?.customerType);
  const qty = quantity > 0 ? quantity : 1;
  const mrp = Number(product.mrp) || 50;
  const minPrice = product.minimumSellingPrice != null && !isNaN(Number(product.minimumSellingPrice))
    ? Number(product.minimumSellingPrice)
    : (product.dealerPrice != null ? Number(product.dealerPrice) : mrp * 0.9);

  // 1. WHOLESALE DEALER: Automatic calculation based on dealer's individual discount %
  if (custType === 'WHOLESALE_DEALER') {
    const dealerDiscount = buyer?.discountPercent != null ? Number(buyer.discountPercent) : 10;
    const base = mrp;
    const discountAmountPerUnit = (base * dealerDiscount) / 100;
    const finalUnitPrice = parseFloat((base - discountAmountPerUnit).toFixed(2));
    const totalPrice = parseFloat((finalUnitPrice * qty).toFixed(2));

    return {
      productId: product.id,
      productName: product.name,
      customerType: 'WHOLESALE_DEALER',
      mrp,
      minimumSellingPrice: minPrice,
      discountPercent: dealerDiscount,
      basePrice: base,
      unitSellingPrice: finalUnitPrice,
      totalPrice,
      quantity: qty,
      isValid: true
    };
  }

  // 2. SHOP: Enforces Minimum Selling Price rule
  if (custType === 'SHOP') {
    const entered = requestedPrice != null && !isNaN(Number(requestedPrice)) ? Number(requestedPrice) : minPrice;
    const isValid = entered >= minPrice;
    const finalUnitPrice = parseFloat(entered.toFixed(2));
    const totalPrice = parseFloat((finalUnitPrice * qty).toFixed(2));

    return {
      productId: product.id,
      productName: product.name,
      customerType: 'SHOP',
      mrp,
      minimumSellingPrice: minPrice,
      discountPercent: 0,
      basePrice: mrp,
      unitSellingPrice: finalUnitPrice,
      totalPrice,
      quantity: qty,
      isValid,
      validationMessage: isValid ? undefined : `Minimum selling price for this product is ₹${minPrice.toFixed(2)}.`
    };
  }

  // 3. CUSTOMER: Manual price entry directly from user
  const manualPrice = requestedPrice != null && !isNaN(Number(requestedPrice))
    ? Number(requestedPrice)
    : (product.retailPrice != null ? Number(product.retailPrice) : mrp);
  const finalUnitPrice = parseFloat(manualPrice.toFixed(2));
  const totalPrice = parseFloat((finalUnitPrice * qty).toFixed(2));

  return {
    productId: product.id,
    productName: product.name,
    customerType: 'CUSTOMER',
    mrp,
    minimumSellingPrice: minPrice,
    discountPercent: 0,
    basePrice: mrp,
    unitSellingPrice: finalUnitPrice,
    totalPrice,
    quantity: qty,
    isValid: true
  };
};

/**
 * Validates whether entered price meets shop minimum price requirement.
 */
export const validateShopMinimumPrice = (product: PricingProduct, price: number): { isValid: boolean; message?: string } => {
  const minPrice = product.minimumSellingPrice != null && !isNaN(Number(product.minimumSellingPrice))
    ? Number(product.minimumSellingPrice)
    : (product.dealerPrice != null ? Number(product.dealerPrice) : (product.mrp || 50) * 0.9);

  if (price < minPrice) {
    return {
      isValid: false,
      message: `Minimum selling price for this product is ₹${minPrice.toFixed(2)}.`
    };
  }
  return { isValid: true };
};
